src/NHibernate/Hql/Ast/ANTLR/QueryTranslatorImpl.cs

```diff
@@ -280,30 +281,25 @@ public string[] ReturnAliases
 			}
 		}
 
-		public bool ContainsCollectionFetches
-		{
-			get
-			{
-				return CollectionFetches.Count > 0;
-			}
-		}
+		public bool ContainsCollectionFetches => CollectionFetches.Count > 0;
 
-		public bool CanAddFetchedCollectionToCache
+		public ISet<ICollectionPersister> UncacheableCollectionPersisters
 		{
 			get
 			{
+				var result = new HashSet<ICollectionPersister>();
```

Comment:
It seems `UncacheableCollectionPersisters` and `ContainsRestrictionOnTable` can be merged:
```
//Find all restrictions in Where clause for cached fetched collections
return new HashSet<ICollectionPersister>(
	ASTUtil.IterateChildrenOfType<FromReferenceNode>(
		((QueryNode) _sqlAst).WhereClause,
		skipSearchInChildrenWhen: node => node.FromElement != null)
	.Select(rn => rn.FromElement)
	.Where(fr => fr?.IsFetch == true && fr.QueryableCollection?.HasCache == true)
	.Select(fr => fr.QueryableCollection)
	);
```

Reply:
Ok, I've applied your suggestion (it got me some time to understand the logic).